### ISSUE TITLE
[BUG] Constantly scale down KCM when shoot is waken up just after a hibernation

### ISSUE THREAD

**Issue by vpnachev** [gardener/dependency-watchdog#18]
**Describe the bug:**
Since quite some time I am observing in my local setup that if try to wake-up a shoot cluster that has been just hibernated, it stuck on wait the nodes to join. Further analysis shows that kube-controller-manager is scaled down, so there is no controller to approve the CSR sent by the node and then track the node status.

If I try to manually scale-up the KCM, after a few seconds it is again scaled down,

```
3m15s       Normal    SuccessfulDelete               replicaset/kube-controller-manager-5f44699448                     Deleted pod: kube-controller-manager-5f44699448-222cf
3m9s        Normal    SuccessfulCreate               replicaset/kube-controller-manager-5f44699448                     Created pod: kube-controller-manager-5f44699448-zwn6s
2m45s       Normal    SuccessfulDelete               replicaset/kube-controller-manager-5f44699448                     Deleted pod: kube-controller-manager-5f44699448-zwn6s
116s        Normal    SuccessfulCreate               replicaset/kube-controller-manager-5f44699448                     Created pod: kube-controller-manager-5f44699448-vvqm2
105s        Normal    SuccessfulDelete               replicaset/kube-controller-manager-5f44699448                     Deleted pod: kube-controller-manager-5f44699448-vvqm2
96s         Normal    SuccessfulCreate               replicaset/kube-controller-manager-5f44699448                     Created pod: kube-controller-manager-5f44699448-s9hst
75s         Normal    SuccessfulDelete               replicaset/kube-controller-manager-5f44699448                     Deleted pod: kube-controller-manager-5f44699448-s9hst
50s         Normal    SuccessfulCreate               replicaset/kube-controller-manager-5f44699448                     Created pod: kube-controller-manager-5f44699448-jqsnq
45s         Normal    SuccessfulDelete               replicaset/kube-controller-manager-5f44699448                     Deleted pod: kube-controller-manager-5f44699448-jqsnq
50s         Normal    ScalingReplicaSet              deployment/kube-controller-manager                                Scaled up replica set kube-controller-manager-5f44699448 to 1
45s         Normal    ScalingReplicaSet              deployment/kube-controller-manager                                Scaled down replica set kube-controller-manager-5f44699448 to 0
```

Before and during the hibernation the kube-apiserver is working just as expected, during the wake up it also starts from the first time - so it is unexpected to me to see DWD scaling down the KCM. Eventually after 10 to 20 minutes of failures, the KCM will be let to scale up.

**Expected behaviour:**
To let the Gardener wake-up hibernated shoot without unnecessary scaling down of KCM.

**How To Reproduce (as minimally and precisely as possible):**
1. Create a shoot.
2. Hibernates the shoot.
3. Wait a minute or two.
4. Wake up the shoot.
5. See the shoot reconciliation is failing with
```yaml
conditions:
  - lastTransitionTime: '2020-06-01T21:46:27Z'
    lastUpdateTime: '2020-06-01T22:27:04Z'
    message: Shoot cluster has been hibernated.
    reason: ConditionNotChecked
    status: 'True'
    type: APIServerAvailable
...
lastOperation:
  description: 'Flow "Shoot cluster reconciliation" encountered task errors: [task
    "Waiting until shoot worker nodes have been reconciled" failed: Error while waiting
    for Worker shoot--li311840--local-aws/local-aws to become ready: extension encountered
    error during reconciliation: Error reconciling worker: Failed while waiting for
    all machine deployments to be ready: ''timed out waiting for the condition'']
    Operation will be retried.'
  lastUpdateTime: '2020-06-01T22:27:31Z'
  progress: 85
  state: Error
  type: Reconcile
```

**Logs:**

#### `dependency-watchdog-probe`
```
I0601 22:19:19.759181       1 scaler.go:219] Starting probes for namespace: shoot--foo--bar
I0601 22:19:19.759219       1 scaler.go:238] Starting the probe in the namespace shoot--foo--bar: &{shoot-kube-apiserver 0xc00007b0c0 [0xc00007b080]}
I0601 22:19:19.759394       1 multicontext.go:60] Registering the context for the key: shoot--foo--bar
I0601 22:19:24.959749       1 prober.go:243] shoot-kube-apiserver/shoot--foo--bar/external: probe failed with error: Get https://api.bar.foo.internal.dev.k8s.ondemand.com/version?timeout=10s: context deadline exceeded (Client.Timeout exceeded while awaiting headers). Will retry...
I0601 22:19:24.959778       1 prober.go:255] shoot-kube-apiserver/shoot--foo--bar/external: probe result: &scaler.probeResult{lastError:(*url.Error)(0xc0002fa1b0), resultRun:1}
I0601 22:19:24.961058       1 prober.go:240] shoot-kube-apiserver/shoot--foo--bar/internal: probe succeeded
I0601 22:19:24.961078       1 prober.go:255] shoot-kube-apiserver/shoot--foo--bar/internal: probe result: &scaler.probeResult{lastError:error(nil), resultRun:2}
I0601 22:19:33.100317       1 reflector.go:235] k8s.io/client-go/informers/factory.go:133: forcing resync
I0601 22:19:34.152630       1 reflector.go:235] k8s.io/client-go/informers/factory.go:133: forcing resync
I0601 22:19:34.961368       1 prober.go:243] shoot-kube-apiserver/shoot--foo--bar/external: probe failed with error: Get https://api.bar.foo.internal.dev.k8s.ondemand.com/version?timeout=10s: context deadline exceeded. Will retry...
I0601 22:19:44.961624       1 prober.go:243] shoot-kube-apiserver/shoot--foo--bar/external: probe failed with error: Get https://api.bar.foo.internal.dev.k8s.ondemand.com/version?timeout=10s: context deadline exceeded (Client.Timeout exceeded while awaiting headers). Will retry...
I0601 22:19:50.059506       1 prober.go:240] shoot-kube-apiserver/shoot--foo--bar/internal: probe succeeded
I0601 22:19:50.059531       1 prober.go:255] shoot-kube-apiserver/shoot--foo--bar/internal: probe result: &scaler.probeResult{lastError:error(nil), resultRun:1}
I0601 22:19:54.961916       1 prober.go:243] shoot-kube-apiserver/shoot--foo--bar/external: probe failed with error: Get https://api.bar.foo.internal.dev.k8s.ondemand.com/version?timeout=10s: net/http: request canceled (Client.Timeout exceeded while awaiting headers). Will retry...
I0601 22:19:54.961946       1 prober.go:255] shoot-kube-apiserver/shoot--foo--bar/external: probe result: &scaler.probeResult{lastError:(*url.Error)(0xc00057b920), resultRun:2}
I0601 22:19:54.963340       1 prober.go:240] shoot-kube-apiserver/shoot--foo--bar/internal: probe succeeded
I0601 22:19:54.963358       1 prober.go:255] shoot-kube-apiserver/shoot--foo--bar/internal: probe result: &scaler.probeResult{lastError:error(nil), resultRun:3}
I0601 22:20:00.059849       1 prober.go:243] shoot-kube-apiserver/shoot--foo--bar/external: probe failed with error: Get https://api.bar.foo.internal.dev.k8s.ondemand.com/version?timeout=10s: context deadline exceeded (Client.Timeout exceeded while awaiting headers). Will retry...
I0601 22:20:03.100636       1 reflector.go:235] k8s.io/client-go/informers/factory.go:133: forcing resync
I0601 22:20:04.153351       1 reflector.go:235] k8s.io/client-go/informers/factory.go:133: forcing resync
I0601 22:20:04.963660       1 prober.go:243] shoot-kube-apiserver/shoot--foo--bar/external: probe failed with error: Get https://api.bar.foo.internal.dev.k8s.ondemand.com/version?timeout=10s: net/http: request canceled (Client.Timeout exceeded while awaiting headers). Will retry...
I0601 22:20:10.060187       1 prober.go:243] shoot-kube-apiserver/shoot--foo--bar/external: probe failed with error: Get https://api.bar.foo.internal.dev.k8s.ondemand.com/version?timeout=10s: context deadline exceeded (Client.Timeout exceeded while awaiting headers). Will retry...
I0601 22:20:14.964020       1 prober.go:243] shoot-kube-apiserver/shoot--foo--bar/external: probe failed with error: Get https://api.bar.foo.internal.dev.k8s.ondemand.com/version?timeout=10s: context deadline exceeded (Client.Timeout exceeded while awaiting headers). Will retry...
I0601 22:20:20.060516       1 prober.go:243] shoot-kube-apiserver/shoot--foo--bar/external: probe failed with error: Get https://api.bar.foo.internal.dev.k8s.ondemand.com/version?timeout=10s: context deadline exceeded (Client.Timeout exceeded while awaiting headers). Will retry...
I0601 22:20:20.060545       1 prober.go:255] shoot-kube-apiserver/shoot--foo--bar/external: probe result: &scaler.probeResult{lastError:(*url.Error)(0xc00089e9c0), resultRun:1}
I0601 22:20:20.061664       1 prober.go:240] shoot-kube-apiserver/shoot--foo--bar/internal: probe succeeded
I0601 22:20:20.061688       1 prober.go:255] shoot-kube-apiserver/shoot--foo--bar/internal: probe result: &scaler.probeResult{lastError:error(nil), resultRun:2}
I0601 22:20:24.356736       1 multicontext.go:63] Unregistering the context for the key: shoot--foo--bar
I0601 22:20:24.356763       1 multicontext.go:68] Cancelling older context for the key: shoot--foo--bar
```

**Screenshots (if applicable):**

**Anything else we need to know?**:
Version: `0.4.0`

**Comment by amshuman-kr** [comment]
@vpnachev Thanks for reporting this. I will try to reproduce the issue.

**Comment by ialidzhikov** [comment]
@rfranzke, this sounds like the issue you were sharing with me today?

**Comment by rfranzke** [comment]
@ialidzhikov No, in my case the KCM was running completely fine - it simply didn't react on CSRs created by new kubelets that tried to join the cluster.

**Comment by ialidzhikov** [comment]
I also hit this issue in my local setup with `v0.4.1`.

**Comment by mvladev** [comment]
The problem is this 

https://github.com/gardener/dependency-watchdog/blob/fc3bd3dd3972024d9a52f01e76f85e3aa4d9e6a0/pkg/scaler/prober.go#L255-L262

It tries to execute 3 times `client.Discovery().ServerVersion()` without any backoff period nor waiting

https://github.com/gardener/dependency-watchdog/blob/fc3bd3dd3972024d9a52f01e76f85e3aa4d9e6a0/pkg/scaler/prober.go#L266-L286

**Comment by amshuman-kr** [comment]
@mvladev But this is only one iteration of the probe. It takes [3](https://github.com/gardener/dependency-watchdog/blob/fc3bd3dd3972024d9a52f01e76f85e3aa4d9e6a0/pkg/scaler/prober.go#L49) or more [consecutive](https://github.com/gardener/dependency-watchdog/blob/fc3bd3dd3972024d9a52f01e76f85e3aa4d9e6a0/pkg/scaler/prober.go#L277) [failed](https://github.com/gardener/dependency-watchdog/blob/fc3bd3dd3972024d9a52f01e76f85e3aa4d9e6a0/pkg/scaler/prober.go#L217) iterations. And each iteration happens at a period of [`30s`](https://github.com/gardener/gardener/blob/master/charts/seed-bootstrap/charts/dependency-watchdog/templates/probe-configmap.yaml#L18).

I am not saying that a delay between the retries won't help. Just pointing out that there is in fact another level of retries and delays on top of this. The tight loop of retries was just to avoid some transient errors)

**Comment by amshuman-kr** [comment]
@vpnachev @mvladev @ialidzhikov I tried to reproduce this issue multiple times but the cluster woke up normally each time. Is it more reproducible with https://github.com/gardener/gardener/pull/2406?

**Comment by vpnachev** [comment]
I am struggling since several months and I have never run Gardener with the SNI feature.

**Comment by rfranzke** [comment]
I just saw that DWD has scaled up KCM when a shoot was hibernated.

**Comment by amshuman-kr** [comment]
> I just saw that DWD has scaled up KCM when a shoot was hibernated.

The watchdog doesn't understand hibernation. It just looks if the shoot apiserver probe succeeds (internally and externally). If the shoot probe succeeds, it would scale up KCM. Not sure why the probe succeeded. That would mean apiserver was up even when the cluster status was saying hibernated.

**Comment by amshuman-kr** [comment]
It looks like gardener scaled down KCM first and apiserver next. It might be a good idea to scale down apiserver first, wait until it is scaled down fully and only then scale down KCM. Otherwise, there will be window where KCM is scaled down but apiserver is still running where the watchdog might interfer.

https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/controlplane.go#L322-L331

**Comment by rfranzke** [comment]
This is because the normal reconciliation flow is used for the hibernation/wake-up as well. The kube-apiserver is usually deployed before any other component that needs to talk to it.
Can't the DWD simply respect the hibernation setting and do nothing in this case?

**Comment by amshuman-kr** [comment]
> Can't the DWD simply respect the hibernation setting and do nothing in this case?

Yes. That makes sense. I merely stating what was happening. To check if the cluster is hibernated, should it check the `Cluster` resource?

**Comment by rfranzke** [comment]
Hm, yes, I think that's the only place where to get the information.

**Comment by amshuman-kr** [comment]
Fixed with #23
